src/VecSim/algorithms/hnsw/hnswlib.h

```diff
@@ -634,6 +635,8 @@ HierarchicalNSW<dist_t>::HierarchicalNSW(SpaceInterface<dist_t> *s, size_t max_e
     max_elements_ = max_elements;
     M_ = M;
     maxM_ = M_;
+    if (M_ > SIZE_MAX / 2)
+        throw std::runtime_error("HNSW index parameter M is too large: argument overflow");
```

Comment:
please do the input validation tests on `M` and not `M_`, before the assignment.